src/sentry/snuba/metrics/query.py

```diff
@@ -286,6 +293,15 @@ def validate_end(self) -> None:
             raise InvalidParams("start must be before end")
 
     def validate_granularity(self) -> None:
+        # Logic specific to how we handle time series in discover
+        if self.use_case_key == UseCaseKey.PERFORMANCE and self.include_series:
+            if self.interval is not None:
+                if self.granularity.granularity > self.interval:
+                    for granularity in METRICS_LAYER_GRANULARITIES:
+                        if granularity < self.interval:
+                            object.__setattr__(self, "granularity", Granularity(granularity))
```

Comment:
Neither of those would work actually because it is a frozen data class
```
dataclasses.FrozenInstanceError: cannot assign to field 'granularity'
```

Reply:
okey let's discuss from another perspective - you are using @frozen dataclass and trying to modify it 🤔 ?

Reply:
I am only modifying it in `__post_init__` in a very specific case, I don't want it to be mutated otherwise.

Reply:
If you still think it's only one way to do it, please add comment why you are doing it. It's not clear. Cause you do modifications inside `validate_granularity`.

Reply:
I can add a comment

Reply:
so may be extract this part into more explicit methd:
```
__update_granularity_special_case(self, **)
```

Reply:
also you tagged a person in your previous comments into this thread, please remove the tag

Reply:
removed :)